vowpalwabbit/conditional_contextual_bandit.cc

```diff
@@ -163,6 +168,36 @@ void inject_slot_features(example* shared, example* slot)
   }
 }
 
+void inject_slot_id(ccb& data, example* shared, int id)
+{
+  // id is zero based, so the vector must be of size id + 1
+  if(id + 1 > data.hashes.size())
+  {
+    data.hashes.resize(id + 1, 0);
+  }
+
+  uint64_t index;
+  if(data.hashes[id] == 0)
+  {
+    auto current_index_str = "index"+std::to_string(id);
+    index = VW::hash_feature(*data.all, current_index_str, data.id_namespace_hash);
+    data.hashes[id] = index;
+  }
+  else
+  {
+    index = data.hashes[id];
+  }
+
+  shared->feature_space[ccb_id_namespace].push_back(1., index);
+}
+
+// Since the slot id is the only thing in this namespace, the popping the value off will work correctly.
+void remove_slot_id(example* shared)
+{
+  shared->feature_space[ccb_id_namespace].indicies.pop();
+  shared->feature_space[ccb_id_namespace].values.pop();
+}
```

Comment:
It's probably worthwhile to also handle the --audit case both (potentially) when adding the feature and (potentially) when removing it.   People need to be able to understand what is happening...

Reply:
Yes audit is definitely worth it, I'll work out how to add that.

Reply:
The trick is adding it in a way which does not computationally impact the audit-free pipeline.  On the upside memory allocation is fine in the audit path :-)

Reply:
I tried adding audit, but it seems I can't get the printout to actually occur. can you take a look at the most recent commit?

Reply:
I don't see a bug here.  Is shared_features_merger working properly with audit?

Reply:
I get audit output for action features that are merged with the shared example so I assume not?

Reply:
Is it because the value of the namespace is 140

Reply:
It should not matter what the namespace is.

Reply:
Maybe use debugger/print? (Is namespace 140 somehow being lost from indices?)

Reply:
The entire feature_space is lost in the stack somewhere, it's going to take me longer to work out what's going on here

Reply:
Good to find this now, obviously...

Reply:
The issue was that I didn't append to the example indices, so when the shared example was merged the feature_space was skipped. I've pushed a fix